sdk/communication/Azure.Communication.Administration/src/Models/PhoneNumberReservationOperation.cs

```diff
@@ -0,0 +1,176 @@
+
+        /// <inheritdocs />
+        public override bool HasCompleted => _hasCompleted;
+
+        /// <inheritdocs />
+        public override bool HasValue => _value != null;
+
+        /// <inheritdocs />
+        public override Response GetRawResponse() => HasCompleted ? _finalRawResponse : _rawResponse;
+
+        /// <summary>
+        /// Check for the latest status of the operation.
+        /// </summary>
+        /// <param name="cancellationToken">
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <returns>The <see cref="Response"/> with the status update.</returns>
+        public override Response UpdateStatus(CancellationToken cancellationToken = default)
+            => UpdateStatusAsync(false, cancellationToken).EnsureCompleted();
+
+        /// <summary>
+        /// Check for the latest status of the operation.
+        /// </summary>
+        /// <param name="cancellationToken">
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <returns>The <see cref="Response"/> with the status update.</returns>
+        public override async ValueTask<Response> UpdateStatusAsync(CancellationToken cancellationToken = default)
+            => await UpdateStatusAsync(true, cancellationToken).ConfigureAwait(false);
+
+        /// <summary>
+        /// Check for the latest status of the operation.
+        /// </summary>
+        /// <param name="cancellationToken">
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <param name="async" />
+        /// <returns>The <see cref="Response"/> with the status update.</returns>
+        private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
+        {
+            using DiagnosticScope scope = _client.ClientDiagnostics.CreateScope($"{nameof(PhoneNumberReservationOperation)}.{nameof(UpdateStatus)}");
+            scope.Start();
+
+            try
+            {
+                if (HasCompleted)
+                {
+                    return GetRawResponse();
+                }
+
+                if (cancellationToken == default)
+                {
+                    cancellationToken = _cancellationToken;
+                }
+
+                Response<PhoneNumberReservation> update = async
+                    ? await _client.GetReservationByIdAsync(reservationId: Id, cancellationToken: cancellationToken).ConfigureAwait(false)
+                    : _client.GetReservationById(reservationId: Id, cancellationToken: cancellationToken);
+
+                var response = update.GetRawResponse();
+                _rawResponse = response;
```

Comment:
If we only set the response when we haven't completed yet, we prevent the race condition of a non-terminal status overriding a terminal status.

```suggestion
          if (!HasCompleted)
          {
                _rawResponse = response;
          }
```

We probably can't guarantee that an early Update never overrides a later Update, unless we number the Update requests or order the Status enum.
For example, "Pending" could override "InProgress" and it's hard to avoid this, a simple lock isnt' enough, we would need ordering.